### PR TITLE
Optimize branch creation

### LINKED ISSUE #2054: Branch creation waits for compaction to finish
Then I tried to understand why GC takes so long time, although it is expected to be cheap operation (reading only layers metadata, do not read or write layer file content - just delete deteriorated layers).  And the reason is compaction_cs which is obtained not only by compaction but also by GC. And compaction can really take very long time. Is it acceptabe to delay creation of new branches and GC until compaction is complete?

_Originally posted by @knizhnik in https://github.com/neondatabase/neon/issues/2053#issuecomment-1178958247_

### PR DESCRIPTION
**Pull request by aome510**
Resolves #2054

**Context**: branch creation needs to wait for GC to acquire `gc_cs` lock, which prevents creating new timelines during GC. However, because individual timeline GC iteration also requires `compaction_cs` lock, branch creation may also need to wait for compactions of multiple timelines. This results in large latency when creating a new branch, which we advertised as *"instantly"*.

This PR optimizes the latency of branch creation by separating GC into two phases:
1. Collect GC data (branching points, cutoff LSNs, etc)
2. Perform GC for each timeline

The GC bottleneck comes from step 2, which must wait for compaction of multiple timelines. This PR modifies the branch creation and GC functions to allow GC to hold the GC lock only in step 1. As a result, branch creation doesn't need to wait for compaction to finish but only needs to wait for GC data collection step, which is fast.

### CODE REVIEW

**Comment by aome510** [comment]
The speedup can be measured by running the `test_branch_creation_heavy_write` test against the PR and `main`. At least based on runs in my local laptop, the maximum latency for this PR should be 3-5x smaller than what of `main`.

**Comment by hlinnaka** [comment]
This doesn't look safe to me. This can happen:

In the beginning, there is only one branch, 'main'. Thread A calls `branch_timeline('main', 'new', '1/11112222)` to create new branch called 'new', at LSN '1/10000000`. At the same time, thread B calls gc_iteration_internal().

1. A acquires layer_removal_cs
2. A reads latest_gc_cutoff_lsn as 1/11110000.
3. B scans the list of timelines. The new timeline creation hasn't finished yet, so it only sees 'main'
4. B calls gc() on the source timeline. It blocks on layer_removal_cs
5. A finishes creating the new timeline
6. B goes ahead with the garbage collection, updates latest_gc_cutoff_lsn to 1/22220000, and removes layers older than that.

Step 6 will remove data that is still needed by the new timeline.

In a nutshell, GC needs to be careful to not remove data that is still needed by child branches. There's a race condition between new branch creation and GC. If a new branch is created after GC has collected its list of branches and their branch-points (`retain_lsns`), but before it has updated `latest_gc_cutoff_lsn` on the source timeline, it can remove data that is still needed by the new timeline.

File: pageserver/src/layered_repository.rs [hunk 1]
```diff
@@ -883,41 +874,42 @@ impl LayeredRepository {
         let mut totals: GcResult = Default::default();
         let now = Instant::now();
 
-        // grab mutex to prevent new timelines from being created here.
-        let _gc_cs = self.gc_cs.lock().unwrap();
-
```
Comment (hlinnaka) [inline pageserver/src/layered_repository.rs:888]: This is the part that prevented the race condition I mentioned in my comment.

**Comment by aome510** [comment]
> This doesn't look safe to me. This can happen:
> 
> In the beginning, there is only one branch, 'main'. Thread A calls `branch_timeline('main', 'new', '1/11112222)` to create new branch called 'new', at LSN '1/10000000`. At the same time, thread B calls gc_iteration_internal().
> 
> 1. A acquires layer_removal_cs
> 2. A reads latest_gc_cutoff_lsn as 1/11110000.
> 3. B scans the list of timelines. The new timeline creation hasn't finished yet, so it only sees 'main'
> 4. B calls gc() on the source timeline. It blocks on layer_removal_cs
> 5. A finishes creating the new timeline
> 6. B goes ahead with the garbage collection, updates latest_gc_cutoff_lsn to 1/22220000, and removes layers older than that.
> 
> Step 6 will remove data that is still needed by the new timeline.
> 
> In a nutshell, GC needs to be careful to not remove data that is still needed by child branches. There's a race condition between new branch creation and GC. If a new branch is created after GC has collected its list of branches and their branch-points (`retain_lsns`), but before it has updated `latest_gc_cutoff_lsn` on the source timeline, it can remove data that is still needed by the new timeline.

Thanks for taking a look into this. I have separated the GC codes into 2 phases as you described in the later comment. Also added a test `test_branch_creation_before_gc` simulating the above scenario.

Still need to do some cleanups such as adding more comments and updating the documents

File: pageserver/src/layered_repository.rs [hunk 2]
```diff
@@ -2340,22 +2353,47 @@ impl LayeredTimeline {
         } else if cfg!(test) {
             // We don't have local timeline in mocked cargo tests.
             // So, just ignore pitr_interval setting in this case.
-            pitr_cutoff_lsn = cutoff;
+            pitr_cutoff_lsn = gc_info.horizon_cutoff;
         }
+        gc_info.pitr_cutoff = pitr_cutoff_lsn;
+
```
Comment (hlinnaka) [inline pageserver/src/layered_repository.rs:2359]: I think you need to also update the timeline's `latest_gc_cutoff_lsn` here. Otherwise essentially the same race condition can still happen:

1. B starts GC and scans the list of timelines. The new timeline doesn't exist yet, so it only sees 'main'
2. B calls update_gc_info() on the source timeline, setting pitr/horizon_cutoff to 1/22220000
3. B releases gc_cs
4. A acquires gc_cs
5. A reads latest_gc_cutoff_lsn as 1/11110000.
6. A finishes creating the new timeline
7. B runs second phase of GC, updates latest_gc_cutoff_lsn to 1/22220000, and removes layers older than that.

Comment (aome510) [reply pageserver/src/layered_repository.rs:2359]: I don't update the gc cutoff there because it's possible to not GC even after updating GC info. For example, that can happen when we go a pageserver shutdown request: https://github.com/neondatabase/neon/pull/2101/files#diff-2dee9987054422607c1ab7369e102927d8477dea6d75c4c7f3891e3d65b15d1eR971-R975

The situation above should not happen because I also added a check for GC info in `branch_timeline`:
https://github.com/neondatabase/neon/pull/2101/files#diff-2dee9987054422607c1ab7369e102927d8477dea6d75c4c7f3891e3d65b15d1eR306-R312

Also the new test simulates that situation, so it should fail if there is a race condition.

File: .github/workflows/build_and_test.yml [hunk 3]
```diff
@@ -110,8 +110,8 @@ jobs:
             target/
           # Fall back to older versions of the key, if no cache for current Cargo.lock was found
           key: |
-            v2-${{ runner.os }}-${{ matrix.build_type }}-cargo-${{ matrix.rust_toolchain }}-${{ hashFiles('Cargo.lock') }}
-            v2-${{ runner.os }}-${{ matrix.build_type }}-cargo-${{ matrix.rust_toolchain }}-
+            v3-${{ runner.os }}-${{ matrix.build_type }}-cargo-${{ matrix.rust_toolchain }}-${{ hashFiles('Cargo.lock') }}
```
Comment (aome510) [inline .github/workflows/build_and_test.yml:113]: Update the cache key to fix the incremental compilation error. See https://neondb.slack.com/archives/C0277TKAJCA/p1658169404632249

File: .github/workflows/codestyle.yml [hunk 4]
```diff
@@ -101,7 +101,7 @@ jobs:
             !~/.cargo/registry/src
             ~/.cargo/git
             target
-          key: ${{ runner.os }}-cargo-${{ hashFiles('./Cargo.lock') }}-rust-${{ matrix.rust_toolchain }}
+          key: v1-${{ runner.os }}-cargo-${{ hashFiles('./Cargo.lock') }}-rust-${{ matrix.rust_toolchain }}
```
Comment (aome510) [inline .github/workflows/codestyle.yml:104]: Update cache key to fix the incremental compilation error.

File: pageserver/src/layered_repository.rs [hunk 5]
```diff
@@ -1080,12 +1120,14 @@ struct GcInfo {
...
-    /// In addition to 'retain_lsns', keep everything newer than 'SystemTime::now()'
-    /// minus 'pitr_interval'
+    /// In addition to 'retain_lsns' and 'horizon_cutoff', keep everything newer than this
+    /// point.
     ///
-    pitr: Duration,
+    /// This is calculated by finding a number such that a record is needed for PITR
+    /// if only if its LSN is larger than 'pitr_cutoff'.
+    pitr_cutoff: Lsn,
 }
```
Comment (hlinnaka) [inline pageserver/src/layered_repository.rs:1131]: You can combine `horizon_cutoff` and `pitr_cutoff` into one value. The code in `gc` prints different debug messages, and uses separate counters depending on which one was smaller, but I don't think that distinction is needed. Instead, you could just print a debug message in `update_gc_info`, to indicate which was dominant.

**Review by hlinnaka — APPROVED** [review]

File: pageserver/src/layered_repository.rs [hunk 6]
```diff
@@ -281,12 +281,21 @@ impl Repository for LayeredRepository {
...
         let _gc_cs = self.gc_cs.lock().unwrap();
 
+        // Besides GC lock, branch creation task doesn't need to hold the `layer_removal_cs` lock,
+        // hence doesn't need to wait for compaction/GC because it ensures that the starting LSN
+        // of the child branch is not out of scope in the middle of the creation task by
+        // 1. holding the GC lock to prevent overwritting timeline GC data
+        // 2. checking both the latest GC cutoff LSN and latest GC info of the source timeline
+        // to avoid initializing the new branch using data removed by past GC iterations
+        // or in-queue GC iterations.
+
```
Comment (hlinnaka) [inline pageserver/src/layered_repository.rs:291]: It's a bit weird to explain in great detail why this *doesn't* need to hold `layer_removal_cs` lock. Also, this seems like a run-on sentence.

File: .github/workflows/build_and_test.yml (hunk 3 above)
Comment (aome510) [reply .github/workflows/build_and_test.yml:113]: I guess it's safe to revert the CI changes for now.

File: pageserver/src/layered_repository.rs (hunk 5 above)
Comment (aome510) [reply pageserver/src/layered_repository.rs:1131]: Yeah. However, it's not a trivial change though. I think it's better to leave that to a separate PR/ follow-up

**Comment by aome510** [comment]
Result when running the new branch creation perf tests comparing with the latest `main`:

This patch:
```
test_branch_creation_heavy_write[20].branch_creation_duration_max: 3.961 s
test_branch_creation_heavy_write[20].branch_creation_duration_avg: 1.185 s
test_branch_creation_heavy_write[20].branch_creation_duration_stdev: 1.222 s
test_branch_creation_many[1024].branch_creation_duration_max: 0.273 s
test_branch_creation_many[1024].branch_creation_duration_avg: 0.102 s
test_branch_creation_many[1024].branch_creation_duration_stdev: 0.037 s
```

main:
```
test_branch_creation_heavy_write[20].branch_creation_duration_max: 21.256 s
test_branch_creation_heavy_write[20].branch_creation_duration_avg: 2.072 s
test_branch_creation_heavy_write[20].branch_creation_duration_stdev: 5.348 s
test_branch_creation_many[1024].branch_creation_duration_max: 3.733 s
test_branch_creation_many[1024].branch_creation_duration_avg: 0.106 s
test_branch_creation_many[1024].branch_creation_duration_stdev: 0.119 s
```

In general, looks like a decent improvement. Will merge once tests passed
